### PR TITLE
Fix `clippy::correctness` in the library

### PR DESCRIPTION
**Pull request by Noratrieb**
needs https://github.com/rust-lang/backtrace-rs/pull/579 to be complete

for https://github.com/rust-lang/compiler-team/issues/709

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Hello Libs team! We want to enable clippy::correctness in CI because this catches bugs (and would have caught at least one bug before) (rust-lang/compiler-team#709). The easiest way to enforce is is to just pass the flag in CI, which also affects the library. This requires a few fixes seen here and of course the commitment of the libs team to continue working with it. Given that correctness has a very low false positive rate, this seems unlikely to cause problems (of course there are some funny false positives, like suggesting using is_nan in the implementation of is_nan).
Is this okay for you?

**Comment by Amanieu** [comment]
We discussed this in the libs meeting and we have no objections to enabling clippy lints for the standard library. We can always revert later if this turns out to cause issues.

**Comment by Nilstrieb** [comment]
Sounds good! Would appreciate a review :)

File: library/core/src/sync/atomic.rs [hunk 1]
```diff
@@ -211,6 +211,8 @@
 #![cfg_attr(not(target_has_atomic_load_store = "8"), allow(dead_code))]
 #![cfg_attr(not(target_has_atomic_load_store = "8"), allow(unused_imports))]
 #![rustc_diagnostic_item = "atomic_mod"]
+// AtomicPtr intrinsics are unsafe and take pointers, but do not dereference them.
+#![allow(clippy::not_unsafe_ptr_arg_deref)]
```
Comment (Amanieu) [inline library/core/src/sync/atomic.rs:215]: Don't these call intrinsics which are themselves unsafe functions?

Comment (Nilstrieb) [reply library/core/src/sync/atomic.rs:217]: Yes, they are unsafe, which is what clippy is concerned about. I elaborated the comment to explain what exactly is happening.

**Comment by joshtriplett** [comment]
r? libs

**Pull request merged**